provisioner/data.js

```diff
@@ -86,6 +88,110 @@ var WorkerType = base.Entity.configure({
   context: ['provisionerId', 'keyPrefix'],
 });
 
+// We want to add a lastModified field
+WorkerType = WorkerType.configure({
+  version: 2,
+  properties: {
+    // These fields are documented in Version 1 of this Entity
+    workerType: base.Entity.types.String,
+    minCapacity: base.Entity.types.Number,
+    maxCapacity: base.Entity.types.Number,
+    scalingRatio: base.Entity.types.Number,
+    minPrice: base.Entity.types.Number,
+    maxPrice: base.Entity.types.Number,
+    canUseOndemand: base.Entity.types.JSON, // delete this
+    canUseSpot: base.Entity.types.JSON, // delete this
+    instanceTypes: base.Entity.types.JSON,
+    regions: base.Entity.types.JSON,
+    // Store the date of last modification for this entity
+    lastModified: base.Entity.types.Date,
+    // Global base UserData object for overwriting
+    userData: base.Entity.types.JSON,
+    // Global base LaunchSpecification object for overwriting
+    launchSpec: base.Entity.types.JSON,
+    // Global base secrets object for overwriting
+    secrets: base.Entity.types.JSON,
+    // Global base scope list for appending, these are the scopes that
+    // temporary TC credentials will be issued against
+    scopes: base.Entity.types.JSON,
+  },
+  migrate: function (item) {
+    // First, let's set up the static/easy data
+    var newWorker = {
+      workerType: item.workerType,
+      minCapacity: item.minCapacity || 0,
+      maxCapacity: item.maxCapacity,
+      scalingRatio: item.scalingRatio,
+      minPrice: item.minPrice,
+      maxPrice: item.maxPrice,
+      canUseOnDemand: item.canUseOnDemand,
+      canUseSpot: item.canUseSpot,
+      lastModified: new Date(),
+    };
+
+    // We do this three times, lets just stick it into a function
+    function fixUserData (x) {
+      var ud = JSON.parse(new Buffer(x, 'base64').toString());
+      // These keys get overwriten automatically by the provisioner
+      // so we don't really need them to ever be in UserData
+      ['provisionerId', 'workerType', 'capacity'].forEach(y => {
+        delete ud[y];
+      });
+      return ud;
+    }
+
+    // Global User Data
+    if (item.launchSpecification.UserData) {
+      newWorker.userData = fixUserData(item.launchSpecification.UserData);
```

Comment:
So if there is no user data, we just omit the property? shouldn't it be `{}` if there is none... I'm not sure?

If so, it seems that it might be better to make `fixUserData` robust enough to return `{}` when given something that isn't a string.